### ISSUE TITLE
Expected behavior with dask-mpi?

### ISSUE THREAD

**Issue by gvarnavi** [dask/distributed#2829]
Hello,

I'm not sure how much of this is expected behavior, but I've spend a considerable amount try to diagnose it and I'm running short on ideas.

I'm trying to run dask with distributed and dask-mpi on NERSC's cori HPC (Haswell compute nodes, 32 cores/node, 128GB/node).
After requesting 2 nodes, I start a background dask-mpi process allocating the scheduler and 2 worker nodes
``` bash
srun -N 2 -n 3 dask-mpi --nthreads 32 --scheduler-file dask_tmp/scheduler.json --memory-limit 0.9 --local-directory dask_tmp/ &
time python mpi-dask-test.py >> mpi-dask-test.out
```
before running the client script:

``` python
import numpy as np
import dask
import dask.array as da
import pickle
from dask.distributed import Client

client=Client(scheduler_file='dask_tmp/scheduler.json')
print("\n"+"#"*100)
print('{:^100s}'.format("Distributed Versions"))
print(client.get_versions(check=True))
print("#"*100+"\n")

nStates=60000
nVertices=10000
chunkStates=20000
chunkVertices=10000

dmm=da.random.random(size=(nStates,nStates),chunks=(chunkStates,chunkStates))
print("Debugging checkpoint 1.0")

projectedVelocities=da.random.random(size=(nStates,nVertices),chunks=(chunkStates,chunkVerti
ces))
print("Debugging checkpoint 2.0")

normalizingSums=da.dot(dmm.T,projectedVelocities)
print(normalizingSums[0,0].compute())
print("Debugging checkpoint 3.0")

broughtDown=normalizingSums.compute()
print(broughtDown[0,0])
print("Debugging checkpoint 4.0")
print("#"*100+"\n")
```

I need access to the full array in the end (to export using h5py), hence the last `compute()` step.
Depending on the sizes of `nStates`, `nVertices`, `chunkStates`, and `chunkVertices` above I get the following behavior:

- Version outputs for all runs:
```
{'scheduler': {'host': (('python', '3.7.3.final.0'), ('python-bits', 64), ('OS', 'Linux'), (
'OS-release', '4.4.103-6.38_4.0.153-cray_ari_c'), ('machine', 'x86_64'), ('processor', 'x86_
64'), ('byteorder', 'little'), ('LC_ALL', 'C.UTF-8'), ('LANG', 'C.UTF-8'), ('LOCALE', 'en_US
.UTF-8')), 'packages': {'required': (('dask', '2.0.0+27.gea59c670'), ('distributed', '2.0.1+
9.g776bb6b7'), ('msgpack', '0.6.1'), ('cloudpickle', '1.2.1'), ('tornado', '6.0.3'), ('toolz
', '0.9.0')), 'optional': (('numpy', '1.16.4'), ('pandas', None), ('bokeh', None), ('lz4', N
one), ('dask_ml', None), ('blosc', None))}}, 'workers': {}, 'client': {'host': [('python', '
3.7.3.final.0'), ('python-bits', 64), ('OS', 'Linux'), ('OS-release', '4.4.103-6.38_4.0.153-
cray_ari_c'), ('machine', 'x86_64'), ('processor', 'x86_64'), ('byteorder', 'little'), ('LC_
ALL', 'C.UTF-8'), ('LANG', 'C.UTF-8'), ('LOCALE', 'en_US.UTF-8')], 'packages': {'required':
[('dask', '2.0.0+27.gea59c670'), ('distributed', '2.0.1+9.g776bb6b7'), ('msgpack', '0.6.1'),
 ('cloudpickle', '1.2.1'), ('tornado', '6.0.3'), ('toolz', '0.9.0')], 'optional': [('numpy',
 '1.16.4'), ('pandas', None), ('bokeh', None), ('lz4', None), ('dask_ml', None), ('blosc', N
one)]}}}
```
- Initial dask-mpi output for all runs:
```
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:    tcp://10.128.2.48:8786
distributed.scheduler - INFO - Receive client connection: Client-1c956c52-a112-11e9-ba51-0001010002ab
distributed.core - INFO - Starting established connection
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.128.2.48:43518'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.128.2.49:46479'
distributed.core - INFO - Event loop was unresponsive in Scheduler for 2.14s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
/global/homes/g/gvarnavi/.conda/envs/dask-ompi/lib/python3.7/contextlib.py:119: UserWarning: Creating scratch directories is taking a surprisingly long time. This is often due to running workers on a network file system. Consider specifying a local-directory to point workers to write scratch data to a local disk.
  next(self.gen)
distributed.worker - INFO -       Start worker at:    tcp://10.128.2.48:37588
distributed.worker - INFO -          Listening to:        tcp://0.0.0.0:37588
distributed.worker - INFO - Waiting to connect to:     tcp://10.128.2.48:8786
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                         32
distributed.worker - INFO -                Memory:                  121.56 GB
distributed.worker - INFO -       Local Directory: /global/cscratch1/sd/gvarnavi/3.DaskEnvironment/ompi-env/dask_tmp/worker-mokm672w
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register tcp://10.128.2.48:37588
distributed.scheduler - INFO - Starting worker compute stream, tcp://10.128.2.48:37588
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:     tcp://10.128.2.48:8786
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
/global/homes/g/gvarnavi/.conda/envs/dask-ompi/lib/python3.7/contextlib.py:119: UserWarning: Creating scratch directories is taking a surprisingly long time. This is often due to running workers on a network file system. Consider specifying a local-directory to point workers to write scratch data to a local disk.
  next(self.gen)
distributed.worker - INFO -       Start worker at:    tcp://10.128.2.49:39412
distributed.worker - INFO -          Listening to:        tcp://0.0.0.0:39412
distributed.scheduler - INFO - Register tcp://10.128.2.49:39412
distributed.worker - INFO - Waiting to connect to:     tcp://10.128.2.48:8786
distributed.scheduler - INFO - Starting worker compute stream, tcp://10.128.2.49:39412
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.worker - INFO -               Threads:                         32
distributed.worker - INFO -                Memory:                  121.56 GB
distributed.worker - INFO -       Local Directory: /global/cscratch1/sd/gvarnavi/3.DaskEnvironment/ompi-env/dask_tmp/worker-ndbtvzay
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:     tcp://10.128.2.48:8786
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
```

- For small numbers, in the order given above (e.g. 10000, 1000, 5000, 1000):
Calculation proceeds without any problems, with the following timing:
```
real    0m44.983s
user    0m1.684s
sys     0m0.840s
```

- For larger numbers (e.g. 30000,10000,10000,5000):
I start getting a number of these warnings (understandably):
```
distributed.core - INFO - Event loop was unresponsive in Worker for 1.75s.  This is often ca
used by long-running GIL-holding functions or moving large chunks of data. This can cause ti
meouts and instability.
```
however, the trend in these seems unintuitive to me. e.g. decreasing `chunkStates` form 10000 to 5000 roughly triples the number of warnings.

- For a somewhat larger calculation (60000,10000,20000,10000) the timing seems rather excessive and I/O bound:
```
real    14m58.376s
user    0m11.268s
sys     0m16.376s
```

To summarize, the questions are:
1. Why am I getting a `Event loop was unresponsive in Scheduler for 2.14s.` warning before it even spins off the workers? what is it trying to do?
2. Is this local-storage warning expected? `UserWarning: Creating scratch directories is taking a surprisingly long time.` I'm using Cori's scratch system (with peak performance of 700GB/sec)
3. Are these timings expected behavior? If I'm interpreting them correctly, this seems very I/O limited
4. What is the correct way to interpret the `Event loop was unresponsive` warnings, if not to decrease chunk sizes?

Any help would be much appreciated!

Cheers,
George

**Comment by mrocklin** [comment]
> Why am I getting a Event loop was unresponsive in Scheduler for 2.14s. warning before it even spins off the workers? what is it trying to do?

I don't know.  It could be lots of things.  Dealing with a bogged down NFS is one common example.

> Is this local-storage warning expected? UserWarning: Creating scratch directories is taking a surprisingly long time. I'm using Cori's scratch system (with peak performance of 700GB/sec)

As above, it's not expected, but we don't know what's happening.  We're just letting you know that things aren't running as smoothly as one would like, and making you aware of it.  If you don't have local storage you might just want to turn off spill-to-disk behavior.  See the following docs:

https://docs.dask.org/en/latest/setup/hpc.html#no-local-storage

> Are these timings expected behavior? If I'm interpreting them correctly, this seems very I/O limited

To be honest I haven't looked too deeply at your code or timings (rushing through issues in the early morning before other work starts).  My usual recommendation here is to use the Dask dashboard.  My guess is that if you look at this while doing your comptuation something will start flashing at you and be fairly easy to diagnose after that.  I recommend watching the dashboard walkthrough youtube video.

See the bottom of https://docs.dask.org/en/latest/diagnostics-distributed.html

If you need access to the dashboard many HPC users use ssh tunnelling.  Given the kinds of things you're talking about I suspect that you're already familiar with this approach, but I'll point you towards this Dask-Jobqueue video as well, which uses ssh-tunneling on an HPC system towards the end

https://youtu.be/FXsgmwpRExM

**Comment by gvarnavi** [comment]
Thanks for the prompt reply, and pardon the delay in replying (ssh-tunneling to NERSC's compute nodes wasn't that straight-forward).

First, turning spill-to-disk behavior off and using [NERSC's Burst-Buffer](https://www.nersc.gov/users/computational-systems/cori/burst-buffer/example-batch-scripts/) doesn't seem to help (as in the `local-storage warnings` indeed go away, but no change in timings/other warnings).
This agrees well with the dashboard diagnostics below, which suggest we are compute-limited.

Secondly, in order to get acess to the dashboard, I had to launch the scheduler through a Jupyter-on-compute node using a `dask-scheduler --scheduler-file dask_tmp/scheduler.json &` command.
I then submit a batch job spinning off the workers to connect to it using:

`srun -N 1 -n 1 dask-worker --nprocs 30 --nthreads 1 --scheduler-file dask_tmp/scheduler.json --memory-limit 0.95 --local-directory dask_tmp/ &`
`srun -N 1 -n 1 dask-worker --nprocs 30 --nthreads 1 --scheduler-file dask_tmp/scheduler.json --memory-limit 0.95 --local-directory dask_tmp/ &`

I've attached typical screenshots of the dashboard below (filenames refer to the same `nStates`, `nVertices`, `chunkStates`, and `chunkVertices` variables as above):
To me these suggest that each numpy-call to sum (yellow bars) takes ~30s for the (10000,5000) chunks (first screenshot) and ~6s for the (5000,2000) chunks (second screenshot).

Why is a call to numpy taking this long?
In-fact, if I naively time a larger calculation on my laptop, I obtain much faster timings:
`time python -c "import numpy as np; np.sum(np.random.rand(300000000)"` ~ 3s

Could it be that dask is somehow chunking the array poorly resulting in an element-wise computation? 
I say this because this is the sort of timing I would expect from a list-comprehension type operation..
`time python -c "[x for x in range(300000000)]"` ~ 15s

Thanks again for you time,
George

Dashboard screenshot for (30000,10000,10000,5000):

![dashboard-30-10-10-5](https://user-images.githubusercontent.com/19610677/61841714-a1240b00-ae63-11e9-9815-181ac4d4dda6.png)

Dashboard screenshot for (30000,10000,5000,2000):

![dashboard-30-10-5-2](https://user-images.githubusercontent.com/19610677/61841715-a1bca180-ae63-11e9-9da2-8fa94647366b.png)

**Comment by mrocklin** [comment]
> Could it be that dask is somehow chunking the array poorly resulting in an element-wise computation?

Perhaps.  You should definitely think about chunking.

Another common cause is oversaturation with MKL.  You might want to look at these doc pages:

-  https://docs.dask.org/en/latest/best-practices.html
-  https://docs.dask.org/en/latest/array-best-practices.html
-  https://docs.dask.org/en/latest/understanding-performance.html

**Comment by GenevieveBuckley** [comment]
I hope those suggestions helped @gvarnavi 
I'll close this issue now since there hasn't been much interaction here for a while
